src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2770,14 +2770,33 @@ module Crystal
       false
     end
 
+    private def allowed_type_in_rescue?(type : UnionType) : Bool
+      type.union_types.all? do |subtype|
+        allowed_type_in_rescue? subtype
+      end
+    end
+
+    private def allowed_type_in_rescue?(type : Crystal::Type) : Bool
+      type.implements?(@program.exception) || type.module?
+    end
+
     def visit(node : Rescue)
       if node_types = node.types
         types = node_types.map do |type|
           type.accept self
           instance_type = type.type.instance_type
-          unless instance_type.implements?(@program.exception)
-            type.raise "#{instance_type} is not a subclass of Exception"
+
+          case instance_type
+          when UnionType
+            unless self.allowed_type_in_rescue? instance_type
+              type.raise "Not all union members of #{instance_type} are a module type or subclass of Exception"
+            end
+          else
+            unless self.allowed_type_in_rescue? instance_type
+              type.raise "#{instance_type} is not a module type or subclass of Exception"
+            end
```

Comment:
**suggestion:** I'm wondering if the distinct error messages are really necessary. It would probably suffice to say something like ``#{instance_type} cannot be used for `rescue`. Only subclasses of `Exception` and modules are allowed``.

Reply:
My thinking there is it makes it sound like you can't use a union; whereas the problem would actually be with the members of that union, and not the union itself.

Reply:
Well the union error message also isn't very precise in this regard as it can also include sub unions 🤷

We could add `or unions thereof` to my suggestion to make this more clear.